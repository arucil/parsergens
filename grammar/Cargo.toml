[package]
name = "grammar"
version = "0.1.0"
authors = ["amlo <xplzjwz@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lalrpop-util = "0.19.0"
bit-set = "0.5.2"
indexmap = "1.5.1"
bimap = "0.5.2"
either = "1.6.0"
codespan-reporting = "0.9.5"
bitvec = "0.18.1"

[build-dependencies]
lalrpop = "0.19.0"

[dev-dependencies]
insta = "0.16.1"
pretty_assertions = "0.6.1"
