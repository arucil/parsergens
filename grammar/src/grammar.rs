use std::ops::Range;
use std::borrow::Borrow;
use super::lexer::{Lexer, TokenId};
use crate::{Set, BiMap, Map};

pub use lower::*;

pub mod lower;

#[derive(Debug)]
pub struct Grammar {
  /// productions of the same nonterminals are consecutive.
  pub rules: Vec<Rule>,
  /// starting non-terminals
  pub start_nts: Set<NonterminalId>,
  /// non-terminals
  pub nts: BiMap<NonterminalId, String>,
  pub nt_prods: Map<NonterminalId, Range<usize>>,
  pub lexer: Lexer,
}

#[derive(Debug)]
pub struct Rule {
  pub nt: NonterminalId,
  pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum Item {
  Nonterminal(NonterminalId),
  Token(TokenId),
  Optional(Vec<Item>),
  Many(Vec<Item>),
  Many1(Vec<Item>),
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
pub struct NonterminalId(u32);

impl NonterminalId {
  pub fn id(&self) -> u32 {
    self.0
  }
}

impl Borrow<u32> for NonterminalId {
  fn borrow(&self) -> &u32 {
    &self.0
  }
}

#[derive(Default)]
pub struct NonterminalIdGen(u32);

impl NonterminalIdGen {
  pub fn gen(&mut self) -> NonterminalId {
    let i = self.0;
    self.0 += 1;
    NonterminalId(i)
  }

  pub fn from(start: u32) -> Self {
    Self(start + 1)
  }
}

impl Grammar {
  pub fn lower(self) -> LoweredGrammar {
    lower::lower(self)
  }
}