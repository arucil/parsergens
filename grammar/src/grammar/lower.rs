use std::ops::Range;
use crate::{
  Grammar, NonterminalIdGen, Item, Lexer, TokenId, NonterminalId
};
use crate::{Set, BiMap, Map};

#[derive(Debug)]
pub struct LoweredGrammar {
  pub prods: Vec<Production>,
  pub start_nts: Set<NonterminalId>,
  pub nts: BiMap<NonterminalId, String>,
  pub nt_metas: Map<NonterminalId, LoweredNonterminalMetadata>,
  pub lexer: Option<Lexer>,
  pub tokens: BiMap<TokenId, String>,
  pub user_code: Vec<String>,
}

#[derive(Debug)]
pub struct Production {
  pub nt: NonterminalId,
  pub kind: ProductionKind,
  pub symbols: Vec<Symbol>,
  pub action: Option<String>,
}

#[derive(Debug)]
pub struct LoweredNonterminalMetadata {
  pub range: Range<usize>,
  pub ty: Option<String>,
  pub kind: NonterminalKind,
}

#[derive(Debug, Clone, Copy)]
pub enum NonterminalKind {
  User,
  Repetition,
  Optional,
}

#[derive(Debug, Clone, Copy)]
pub enum ProductionKind {
  Ordinary,
  RepetitionFirst,
  RepetitionRest,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Symbol {
  Nonterminal(NonterminalId),
  Token(TokenId),
}

pub(super) fn lower(grammar: Grammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .left_values()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut lowered = LoweredGrammar {
    prods: vec![],
    start_nts: grammar.start_nts,
    nts: grammar.nts,
    nt_metas: Map::new(),
    lexer: grammar.lexer,
    tokens: grammar.tokens,
    user_code: grammar.user_code,
  };

  for (nt, meta) in grammar.nt_metas {
    let mut rules = vec![];

    for rule in &grammar.rules[meta.range] {
      rules.push((
        lower_items(&rule.items, &mut lowered, &mut nt_id_gen),
        rule.action.clone()
      ));
    }

    let start = lowered.prods.len();

    for (symbols, action) in rules {
      lowered.prods.push(Production {
        nt,
        kind: ProductionKind::Ordinary,
        symbols,
        action,
      });
    }

    let end = lowered.prods.len();

    let meta = LoweredNonterminalMetadata {
      range: start..end,
      ty: meta.ty,
      kind: NonterminalKind::User,
    };

    lowered.nt_metas.insert(nt, meta);
  }

  lowered
}

fn lower_items(
  items: &[Item],
  lowered: &mut LoweredGrammar,
  nt_id_gen: &mut NonterminalIdGen,
) -> Vec<Symbol> {
  items.iter().map(|item| {
    match item {
      Item::Nonterminal(nt) => Symbol::Nonterminal(*nt),
      Item::Token(token) => Symbol::Token(*token),
      Item::Optional(items) => {
        let symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '?');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols: vec![],
          action: None,
        });
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::Ordinary,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Optional,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '*');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: vec![],
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
      Item::Many1(items) => {
        let mut symbols = lower_items(items, lowered, nt_id_gen);
        let start = lowered.prods.len();
        let nt = nt_id_gen.gen();
        let name = make_production_name(&symbols, lowered, '+');
        lowered.nts.insert(nt, name);

        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionFirst,
          symbols: symbols.clone(),
          action: None,
        });

        symbols.insert(0, Symbol::Nonterminal(nt));
        lowered.prods.push(Production {
          nt,
          kind: ProductionKind::RepetitionRest,
          symbols,
          action: None,
        });

        let end = lowered.prods.len();

        lowered.nt_metas.insert(nt, LoweredNonterminalMetadata {
          range: start..end,
          ty: None,
          kind: NonterminalKind::Repetition,
        });

        Symbol::Nonterminal(nt)
      }
    }
  }).collect()
}

fn make_production_name(
  symbols: &[Symbol],
  lowered: &LoweredGrammar,
  suffix: char
) -> String {
  let mut buf = "(".to_owned();
  let mut space = false;

  for symbol in symbols {
    if space {
      buf.push(' ');
    }

    match symbol {
      Symbol::Nonterminal(nt) => {
        buf.push_str(lowered.nts.get_by_left(nt).unwrap());
      }
      Symbol::Token(token) => {
        buf.push_str(lowered.tokens.get_by_left(token).unwrap());
      }
    }

    space = true;
  }

  buf.push(')');
  buf.push(suffix);

  buf
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn lower_repetition() {
    let grammar = crate::build(r#"
%token A "a"

%start top

top = A (atom (A)? )* A
  | (A atom)+

atom = A
  | ()
    "#).unwrap();

    let lowered = lower(grammar);

    assert_debug_snapshot!(lowered);
  }
}