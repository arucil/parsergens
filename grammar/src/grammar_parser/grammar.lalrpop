use lalrpop_util::ParseError;
use super::ast::*;
use super::lex;
use super::UserParseError;
use super::regex;

#[LALR]
grammar<'input>(input: &'input str);

pub Document: Grammar = (Separator <decl>)*
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned((l, r), Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned((l, r), Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned((l, r), Decl::Token(decl))),
  <l:@L> <decl:ext_token_decl> <r:@L> => Some(Spanned((l, r), Decl::ExtToken(decl))),
  <l:@L> <decl:skip_decl> <r:@L> => Some(Spanned((l, r), Decl::Skip(decl))),
  <l:@L> <decl:user_decl> <r:@L> => Some(Spanned((l, r), Decl::User(decl))),
  <l:@L> <decl:assoc_decl> <r:@L> => Some(Spanned((l, r), Decl::Assoc(decl))),
  <l:@L> <decl:state_decl> <r:@L> => Some(Spanned((l, r), Decl::State(decl))),
};

rule_decl: RuleDecl =
  <name:ident> <ty:(Colon <code_block>)?> Assign <alt1:production> <mut alts:(Or <production>)*>
  =>? {
    alts.insert(0, alt1);
    if alts.iter().any(|alt| alt.1.action.is_some()) && ty.is_none() {
      Err(ParseError::User {
        error: UserParseError::MissingType(name.0),
      })
    } else {
      Ok(RuleDecl {
        name,
        ty,
        alts,
      })
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:term+> <prec:(Percent Prec <ident>)?> <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Terms(terms),
        prec,
        action,
      }
    ),
  <l:@L> epsilon <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Epsilon,
        prec: None,
        action,
      }
    ),
};

epsilon: () = LParen RParen;

term: Term = {
  ident => Term::Symbol(<>),
  LParen <terms:term+> RParen QuestionMark => Term::Optional(terms),
  LParen <terms:term+> RParen Asterisk => Term::Many(terms),
  LParen <terms:term+> RParen Plus => Term::Many1(terms),
}

start_decl: StartDecl = Percent Start <name:ident>
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <pattern:token_pattern>
  => {
    TokenDecl {
      name,
      pattern,
    }
  };

ext_token_decl: ExternalTokenDecl = Percent Token <name:ident>
  => {
    ExternalTokenDecl {
      name,
    }
  };

skip_decl: SkipDecl = Percent Skip <pattern:token_pattern>
  => {
    SkipDecl {
      pattern,
    }
  };

user_decl: UserDecl = Percent User <code:code_block>
  => {
    UserDecl {
      code,
    }
  };

assoc_decl: AssocDecl = Percent <assoc:assoc> <names:ident+>
  => {
    AssocDecl {
      names,
      assoc,
    }
  };

assoc: Spanned<Assoc> = {
  <l:@L> LeftAssoc <r:@L> => Spanned((l, r), Assoc::LeftAssoc),
  <l:@L> RightAssoc <r:@L> => Spanned((l, r), Assoc::RightAssoc),
  <l:@L> NonAssoc <r:@L> => Spanned((l, r), Assoc::NonAssoc),
}

state_decl: StateDecl = Percent State <state:code_block>
  => {
    StateDecl {
      state,
    }
  };

token_pattern: TokenPattern = {
  <regex:regex> =>? regex::parse_regex(
    &regex.1[1..regex.1.len() - 1],
    regex.0 .0 + 1
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::Regex,
      source: regex,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error),
    }),
  <str:string> =>? regex::parse_string_literal(
    &str.1[1..str.1.len() - 1], str.0 .0
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::String,
      source: str,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error)
    }),
}

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned((l, r), x.text.to_owned());

ident = spanned<Ident>;
string = spanned<String>;
regex = spanned<Regex>;
code_block: Spanned<String> = <spanned<CodeBlock>> =>
  Spanned(<>.0, <>.1.trim_start_matches('{').trim_end_matches('}').trim().to_owned());

extern {
  type Location = usize;
  type Error = UserParseError;

  enum lex::Token<'input> {
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Skip => lex::Token { kind: lex::TokenKind::Skip, .. },
    User => lex::Token { kind: lex::TokenKind::User, .. },
    LeftAssoc => lex::Token { kind: lex::TokenKind::LeftAssoc, .. },
    RightAssoc => lex::Token { kind: lex::TokenKind::RightAssoc, .. },
    NonAssoc => lex::Token { kind: lex::TokenKind::NonAssoc, .. },
    Prec => lex::Token { kind: lex::TokenKind::Prec, .. },
    State => lex::Token { kind: lex::TokenKind::State, .. },

    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },
    String => lex::Token { kind: lex::TokenKind::String, .. },
    CodeBlock => lex::Token { kind: lex::TokenKind::CodeBlock, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    LParen => lex::Token { kind: lex::TokenKind::LParen, .. },
    RParen => lex::Token { kind: lex::TokenKind::RParen, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },
    Colon => lex::Token { kind: lex::TokenKind::Colon, .. },
    Asterisk => lex::Token { kind: lex::TokenKind::Asterisk, .. },
    QuestionMark => lex::Token { kind: lex::TokenKind::QuestionMark, .. },
    Plus => lex::Token { kind: lex::TokenKind::Plus, .. },
    Arrow => lex::Token { kind: lex::TokenKind::Arrow, .. },

    Enter => lex::Token { kind: lex::TokenKind::Enter, .. },
    Leave => lex::Token { kind: lex::TokenKind::Leave, .. },
    Separator => lex::Token { kind: lex::TokenKind::Separator, .. },
  }
}