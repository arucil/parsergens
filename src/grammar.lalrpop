use super::ast::*;
use super::lex;

#[LALR]
grammar<'input>(input: &'input str);

pub document: Grammar = <decl*>
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned(l..r, Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned(l..r, Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned(l..r, Decl::Token(decl))),
  Newline => None,
};

rule_decl: RuleDecl =
  <name:ident> Assign <alt1:production> Newline <mut alts:(Indent Or <production> Newline)*>
  => {
    alts.insert(0, alt1);
    RuleDecl {
      name,
      alts,
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:ident*> <r:@L> => Spanned(l..r, RuleAlt::Terms(terms)),
  <l:@L> Epsilon <r:@L> => Spanned(l..r, RuleAlt::Epsilon),
};

start_decl: StartDecl = Percent Start <name:ident> Newline
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <regex:regex>
  => {
    TokenDecl {
      name,
      regex,
    }
  };

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned(l..r, x.text.to_owned());

ident = spanned<Ident>;
regex = spanned<Regex>;

extern {
  type Location = usize;
  type Error = lex::LexError;

  enum lex::Token<'input> {
    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    Epsilon => lex::Token { kind: lex::TokenKind::Epsilon, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },

    Indent => lex::Token { kind: lex::TokenKind::Indent, .. },
    Newline => lex::Token { kind: lex::TokenKind::Newline, .. },
  }
}