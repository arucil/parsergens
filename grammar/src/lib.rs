#![feature(or_patterns, never_type)]

use lalrpop_util::ParseError;
use codespan_reporting::diagnostic::{Diagnostic, Label};
use codespan_reporting::term;
use codespan_reporting::term::termcolor::{ColorChoice, StandardStream};
use codespan_reporting::files::SimpleFile;

mod grammar_parser;

pub mod lexer;
pub mod grammar;

pub use crate::grammar::*;
pub use lexer::{Lexer, TokenId, TokenIdGen};

use grammar_parser::ast;
use grammar_parser::lex::{Token, LexErrorKind, LexError};
use grammar_parser::regex::{RegexError, RegexErrorKind};
use grammar_parser::UserParseError;
use lexer::LexerError;

#[cfg(not(debug_assertions))]
use std::collections::{HashMap, HashSet};
#[cfg(not(debug_assertions))]
use bimap::BiHashMap;

#[cfg(debug_assertions)]
use indexmap::{IndexMap, IndexSet};
#[cfg(debug_assertions)]
use bimap::BiBTreeMap;

#[cfg(not(debug_assertions))]
type Map<K, V> = HashMap<K, V>;

#[cfg(debug_assertions)]
type Map<K, V> = IndexMap<K, V>;

#[cfg(not(debug_assertions))]
type BiMap<K, V> = BiHashMap<K, V>;

#[cfg(debug_assertions)]
type BiMap<K, V> = BiBTreeMap<K, V>;

#[cfg(not(debug_assertions))]
type Set<K> = HashSet<K>;

#[cfg(debug_assertions)]
type Set<K> = IndexSet<K>;


#[derive(Debug)]
pub struct GrammarError {
  kind: GrammarErrorKind,
  message: String,
  span: (usize, usize),
}

#[derive(Debug)]
pub enum GrammarErrorKind {
  ParseError,
  SymbolNotFound,
  NameConflict,
}

pub fn build(grammar: &str) -> Result<Grammar, GrammarError> {
  let ast = grammar_parser::parse(grammar)
    .map_err(|err| err.into())?;

  let decls = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Token(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let skips = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Skip(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let starts = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Start(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let rules = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Rule(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let lexer = Lexer::new(&decls, &skips).map_err(|err| {
    match err {
      LexerError::NoTokens => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("no token declarations"),
          span: (0, grammar.len()),
        }
      }
      LexerError::RegexError(err) => {
        err.into()
      }
    }
  })?;

  let mut nt_id_gen = NonterminalIdGen::default();
  let nts = rules.iter()
    .map(|decl| {
      if lexer.tokens.get_by_right(&decl.name.1).is_some() {
        return Err(GrammarError {
          kind: GrammarErrorKind::NameConflict,
          message: format!("rule name collides with token name"),
          span: decl.name.0,
        });
      }

      let id = nt_id_gen.gen();

      Ok((id, decl.name.1.clone()))
    })
    .collect::<Result<BiMap<_, _>, GrammarError>>()?;

  let start_nts = starts.iter()
    .map(|decl| {
      if let Some(&id) = nts.get_by_right(&decl.name.1) {
        Ok(id)
      } else {
        Err(GrammarError {
          kind: GrammarErrorKind::SymbolNotFound,
          message: format!("nonterminal not found"),
          span: decl.name.0,
        })
      }
    })
    .collect::<Result<Set<_>, GrammarError>>()?;

  let mut prods = vec![];
  let mut nt_prods = Map::new();

  for rule in &rules {
    let nt = *nts.get_by_right(&rule.name.1).unwrap();
    let start = prods.len();

    for alt in &rule.alts {
      let items = match &alt.1 {
        ast::RuleAlt::Epsilon => vec![],
        ast::RuleAlt::Terms(terms) => {
          convert_terms(terms, &nts, &lexer.tokens)?
        }
      };

      prods.push(Production {
        nt,
        items,
      });
    }

    nt_prods.insert(nt, start..prods.len());
  }

  Ok(Grammar {
    prods,
    start_nts,
    nts,
    nt_prods,
    lexer,
  })
}

fn convert_terms(
  terms: &[ast::Term],
  nts: &BiMap<NonterminalId, String>,
  tokens: &BiMap<TokenId, String>,
) -> Result<Vec<Item>, GrammarError> {
  let mut items = vec![];

  for term in terms {
    match term {
      ast::Term::Symbol(sym) => {
        if let Some(&id) = nts.get_by_right(&sym.1) {
          items.push(Item::Nonterminal(id));
        } else if let Some(&id) = tokens.get_by_right(&sym.1) {
          items.push(Item::Token(id));
        } else {
          return Err(GrammarError {
            kind: GrammarErrorKind::SymbolNotFound,
            message: format!("nonterminal or token not found"),
            span: sym.0,
          });
        }
      }
      ast::Term::Optional(terms) => {
        items.push(Item::Optional(convert_terms(terms, nts, tokens)?));
      }
      ast::Term::Many(terms) => {
        items.push(Item::Many(convert_terms(terms, nts, tokens)?));
      }
      ast::Term::Many1(terms) => {
        items.push(Item::Many1(convert_terms(terms, nts, tokens)?));
      }
    }
  }

  Ok(items)
}

pub fn report_error(input: &str, error: &GrammarError) {
  let writer = StandardStream::stderr(ColorChoice::Always);
  let config = term::Config::default();
  let files = SimpleFile::new("", input);

  let diagnostic = Diagnostic::error();
  let diagnostic = match error.kind {
    GrammarErrorKind::ParseError => {
      diagnostic.with_message("syntax error")
    }
    GrammarErrorKind::NameConflict => {
      diagnostic.with_message("name conflict")
    }
    GrammarErrorKind::SymbolNotFound => {
      diagnostic.with_message("symbol not found")
    }
  };
  let diagnostic = diagnostic.with_labels(vec![
    Label::primary((), error.span.0..error.span.1).with_message(&error.message)
  ]);

  term::emit(&mut writer.lock(), &config, &files, &diagnostic).unwrap();
}

impl<'a> Into<GrammarError> for ParseError<usize, Token<'a>, UserParseError> {
  fn into(self) -> GrammarError {
    match self {
      Self::InvalidToken { location } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid token or EOF"),
          span: (location, location),
        }
      }
      Self::UnrecognizedEOF { location, expected } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("expected {}, found EOF", expected.join(", ")),
          span: (location, location),
        }
      }
      Self::UnrecognizedToken { token, expected } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("expected {}, found {}", expected.join(", "), token.1),
          span: (token.0, token.2),
        }
      }
      Self::ExtraToken { token } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: token.1.to_string(),
          span: (token.0, token.2),
        }
      }
      Self::User { error } => {
        match error {
          UserParseError::LexError(error) => error.into(),
          UserParseError::RegexError(error) => error.into(),
        }
      }
    }
  }
}

impl<'a> Into<GrammarError> for LexError {
  fn into(self) -> GrammarError {
    match self.kind {
      LexErrorKind::UnclosedRegex => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("unclosed regex"),
          span: self.span,
        }
      }
      LexErrorKind::UnclosedString => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("unclosed string"),
          span: self.span,
        }
      }
      LexErrorKind::InvalidChar => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid character"),
          span: self.span,
        }
      }
    }
  }
}

impl<'a> Into<GrammarError> for RegexError {
  fn into(self) -> GrammarError {
    match self.kind {
      RegexErrorKind::SyntaxError => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("regex syntax error"),
          span: self.span,
        }
      }
      RegexErrorKind::InvalidCodePoint => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid code point"),
          span: self.span,
        }
      }
      RegexErrorKind::Empty => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("regex or string may accept empty string"),
          span: self.span,
        }
      }
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn expr() {
    let input = r#"
%start expr

%token PLUS "+"
%token MINUS "-"
%token MUL "*"
%token DIV "/"
%token LPAREN "("
%token RPAREN ")"
%token COMMA ","
%token NUMBER /\d+(\.\d*)?/
%token IDENT /[a-zA-Z][\w_]*/

%skip /[ \n]/

expr = expr PLUS expr
  | expr MINUS expr
  | expr MUL expr
  | expr DIV expr
  | factor

factor =
    NUMBER
  | LPAREN expr RPAREN
  | MINUS factor
  | IDENT
  | IDENT LPAREN param_list RPAREN

param_list =
    ()
  | expr ( COMMA expr )*
    "#;

    assert_debug_snapshot!(build(input));
  }
}