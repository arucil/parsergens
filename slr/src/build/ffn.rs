//! compute FIRST, FOLLOW, and NULLABLE sets.

use bit_set::BitSet;
use grammar::{Grammar, NonterminalId, Item};
use crate::{Map, Set};

pub struct FfnResult {
  first: Map<NonterminalId, BitSet>,
  follow: Map<NonterminalId, BitSet>,
  nullable: Set<NonterminalId>,
}

pub fn compute(grammar: &Grammar) -> FfnResult {
  let nullable = compute_nullable(grammar);
  todo!()
}

fn compute_nullable(grammar: &Grammar) -> Set<NonterminalId> {
  let mut nullable = Set::new();
  let mut visited = Set::new();
  for &nonterminal in grammar.nonterminals.left_values() {
    compute_nullable_help(grammar, nonterminal, &mut nullable, &mut visited);
  }
  nullable
}

fn compute_nullable_help(
  grammar: &Grammar,
  nonterminal: NonterminalId,
  nullable: &mut Set<NonterminalId>,
  visited: &mut Set<NonterminalId>,
) -> bool {
  if visited.contains(&nonterminal) {
    return nullable.contains(&nonterminal);
  }

  let range = grammar.nonterminal_productions
    .get(&nonterminal).unwrap().clone();

  for production in &grammar.productions[range] {
    if compute_items_nullable_help(grammar, &production.items, nullable, visited) {
      nullable.insert(nonterminal);
      return true;
    }
  }

  false
}

fn compute_items_nullable_help(
  grammar: &Grammar,
  items: &[Item],
  nullable: &mut Set<NonterminalId>,
  visited: &mut Set<NonterminalId>,
) -> bool {
  items.iter().all(|item| {
    match item {
      Item::Nonterminal(nonterminal) => {
        compute_nullable_help(grammar, *nonterminal, nullable, visited)
      }
      Item::Token(_) => false,
      Item::Optional(_) => true,
      Item::Many(_) => true,
      Item::Many1(items) => {
        compute_items_nullable_help(grammar, items, nullable, visited)
      }
    }
  })
}