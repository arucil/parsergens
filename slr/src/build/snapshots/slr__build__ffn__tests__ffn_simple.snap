---
source: slr/src/build/ffn.rs
expression: ffn
---
Ffn {
    first: {
        NonterminalId(
            1,
        ): {
            1,
        },
        NonterminalId(
            2,
        ): {
            0,
            1,
        },
        NonterminalId(
            0,
        ): {
            0,
            1,
            2,
        },
    },
    follow: {
        NonterminalId(
            0,
        ): {},
        NonterminalId(
            1,
        ): {
            0,
            1,
            2,
        },
        NonterminalId(
            2,
        ): {
            0,
            1,
            2,
        },
    },
    nullable: {
        1,
        2,
    },
}
